### PR TITLE
Allow oauth secret rotation

### PR DESCRIPTION
**Pull request by adelowo**
Fixes https://github.com/TykTechnologies/tyk/issues/2460

### CODE REVIEW

**Comment by buger** [comment]
Nice, but this require UI change too.

**Comment by adelowo** [comment]
I will create the ticket on FE repo.

**Comment by letzya** [comment]
@adelowo @dencoded 
1. so we don't enforce any restrictions on the secret?
2. is it possible to also auto-gen a secret and return it? this way the same code that generated the first secret will generate the new one

**Comment by buger** [comment]
> is it possible to also auto-gen a secret and return it? this way the same code that generated the first secret will generate the new one

That's a very good point I actually missed.
Let's not allow specify it manually, and rotate using the existing algorithm.

**Comment by dencoded** [comment]
I think this depends on rotation approach:
1. rotation originating from oauth client when it gets implemented with new secret and then need to update secret on Tyk side (this is I guess suggested in this PR approach)
2. rotation is done 1st on Tyk side (secret gets generated by Tyk) and then oauth client implementation gets updated with new secret generated by Tyk. Here existing oauth clients with old secret will start failing right away as there is no way they know new secret

In approach 1. - I think it is possible to do graceful transition to new secret without down time - i.e. try old one, then try new one if failed.

However, it seems when ever we choose  - it is worth to add on tyk side concept of grace period when old secret can be used for some time after new one gets generated.

**Comment by letzya** [comment]
if tyk will initiate the rotation that means we must notify the admin and add grace period. If we do that, someone needs to do a little research to see if anyone does that. I could find anything on the wild or rfc. OAuth let you rotate manually with a click of a button (auto-gen key!)

I think for this release we need to give api to rotate and tyk should auto-gen it.
If we want to allow users to set it up on their own we must set some policy, like we have for passwords - so I'd rather not allow it for in this release. We should also check what other AS are doing and not do something specific to us

**Comment by adelowo** [comment]
I have updated this to make sure the secret is autogenerated by GW @buger @letzya

**Comment by letzya** [comment]
Shouldn't we split the update of an oauth client from a secret rotation? why does the same function need to do both?

**Comment by adelowo** [comment]
Well, `updateOauthClient` is an internal function and the operations are actually similar except for the `if forceSecret`.. @dencoded thoughts?

**Comment by dencoded** [comment]
I don't see anything bad if two endpoints reuse the same data access logic but with slightly different params, I actually would prefer that.

**Comment by buger** [comment]
Yeah, I like having it as separate endpoint as well. It just re-uses original logic.

File: gateway/api.go [hunk 1]
```diff
@@ -1688,6 +1694,16 @@ func invalidateOauthRefresh(w http.ResponseWriter, r *http.Request) {
 	doJSONWrite(w, http.StatusOK, success)
 }
 
+func rotateOauthClientHandler(w http.ResponseWriter, r *http.Request) {
+
+	apiID := mux.Vars(r)["apiID"]
+	keyName := mux.Vars(r)["keyName"]
+
+	obj, code := updateOauthClient(keyName, apiID, true, r)
```
[the remark this answers is not in the document]
Comment (buger) [reply gateway/api.go]: From REST point of view, every action which "modify" smth anyhow should be either POST or PUT

[the remark this answers is not in the document]
Comment (letzya) [reply gateway/api.go]: the PUT is not the issue here. I'm aware of this "update" but feels that the"/rotate" only set the argument of the method "updateOauthClient" to true with no other updates and the "/update" endpoint updates the other fields of the client so they dont really intersect

**Comment by buger** [comment]
Actually now I get what @letzya means. I missed it.

Indeed if you have a separate endpoint for secret rotation, it should not require having a body with oauth client payload. Just make it PUT request with empty body.

**Comment by adelowo** [comment]
Thanks @letzya .

@buger It has been updated.

File: gateway/server.go [hunk 2]
```diff
@@ -436,6 +436,7 @@ func loadAPIEndpoints(muxer *mux.Router) {
 		r.HandleFunc("/health", healthCheckhandler).Methods("GET")
 		r.HandleFunc("/oauth/clients/create", createOauthClient).Methods("POST")
 		r.HandleFunc("/oauth/clients/{apiID}/{keyName:[^/]*}", oAuthClientHandler).Methods("PUT")
+		r.HandleFunc("/oauth/clients/{apiID}/{keyName:[^/]*}/rotate", rotateOauthClientHandler).Methods("PUT")
```
Comment (letzya) [inline gateway/server.go]: can we change the "keyName" to "ClientId" for clarity?

**Comment by letzya** [comment]
@adelowo have created pr with a suggestion https://github.com/adelowo/tyk/pull/1

**Comment by letzya** [comment]
FYI - haven't reviewed the tests.

**Pull request merged**